pupil_src/shared_modules/video_capture/file_backend.py

```diff
@@ -104,6 +104,43 @@ def gray(self):
         return self.img[:, :, 0]  # return first channel
 
 
+class Buffering:
```

Comment:
I like this abstraction but I think we did not unlock its full potential.

`Buffering.state` is basically a wrapper for `File_Source.video_stream`. Therefore, my suggestion would be modify `File_Source._get_streams()` as follows:

- Abstract `Decoder` class that defines decoding interface: `seek()`, `next_frame()`, `frame_size`
- Rename `Buffer -> BufferedDecoder` and `NotBuffered -> OnDemandDecoder`, both inherting from `Decoder`
- `File_Source._get_streams()` receives an additional argument: `should_buffer`
- It returns a 2-element tuple including an audio stream as before. But instead of an `av.Video_Stream`, it returns a `Decoder` instance (depending on `should_buffer`)

Consequences:
- We do not need `File_Source.buffer` anymore
- All `File_Source.buffer` api calls are called on the (new) `File_Source.video_stream` (Decoder instance)
- `File_Source.frame_size` simply returns `File_Source.video_stream.frame_size`

This way we can hide the `av.Video_Stream` completely in the `Decoder` classes.